File: go-opencode/internal/tool/task.go
package tool

import (
	"context"
	"encoding/json"
	"fmt"

	einotool "github.com/cloudwego/eino/components/tool"
	"github.com/opencode-ai/opencode/internal/agent"
)

const taskDescription = `Launch a new agent to handle complex, multi-step tasks autonomously.

The Task tool launches specialized agents (subprocesses) that autonomously handle complex tasks.
Each agent type has specific capabilities and tools available to it.

Available agent types:
- general: General-purpose agent for researching and exploration
- explore: Fast agent specialized for codebase exploration
- plan: Planning agent for analysis without making changes

Usage notes:
- Launch multiple agents concurrently when possible
- Each agent invocation is stateless
- The agent's outputs should be trusted
- Specify desired thoroughness level when calling explore agent`

// TaskTool allows spawning sub-agents for complex tasks.
type TaskTool struct {
	workDir       string
	agentRegistry *agent.Registry
	executor      TaskExecutor
}

// TaskExecutor is the interface for executing subtasks.
type TaskExecutor interface {
	// ExecuteSubtask runs a subtask with the given agent and prompt.
	ExecuteSubtask(ctx context.Context, sessionID string, agentName string, prompt string, opts TaskOptions) (*TaskResult, error)
}

// TaskOptions contains options for task execution.
type TaskOptions struct {
	Model       string // Optional model override (sonnet, opus, haiku)
	ResumeFrom  string // Optional agent ID to resume from
	Description string // Short description of the task
}

// TaskResult represents the result of a subtask.
type TaskResult struct {
	Output    string         `json:"output"`
	SessionID string         `json:"sessionID"`
	AgentID   string         `json:"agentID,omitempty"`
	Error     string         `json:"error,omitempty"`
	Metadata  map[string]any `json:"metadata,omitempty"`
}

// TaskInput represents the input for the task tool.
// SDK compatible: uses camelCase field names to match TypeScript.
type TaskInput struct {
	Description  string `json:"description"`
	Prompt       string `json:"prompt"`
	SubagentType string `json:"subagentType"`
	Model        string `json:"model,omitempty"`
	Resume       string `json:"resume,omitempty"`
}

// NewTaskTool creates a new task tool.
func NewTaskTool(workDir string, registry *agent.Registry) *TaskTool {
	if registry == nil {
		registry = agent.NewRegistry()
	}
	return &TaskTool{
		workDir:       workDir,
		agentRegistry: registry,
	}
}

// SetExecutor sets the task executor.
func (t *TaskTool) SetExecutor(executor TaskExecutor) {
	t.executor = executor
}

func (t *TaskTool) ID() string          { return "Task" }
func (t *TaskTool) Description() string { return taskDescription }

func (t *TaskTool) Parameters() json.RawMessage {
	return json.RawMessage(`{
		"type": "object",
		"properties": {
			"description": {
				"type": "string",
				"description": "A short (3-5 word) description of the task"
			},
			"prompt": {
				"type": "string",
				"description": "The detailed task for the agent to perform"
			},
			"subagentType": {
				"type": "string",
				"description": "The type of specialized agent to use (general, explore, plan)"
			},
			"model": {
				"type": "string",
				"description": "Optional model to use (sonnet, opus, haiku)",
				"enum": ["sonnet", "opus", "haiku"]
			},
			"resume": {
				"type": "string",
				"description": "Optional agent ID to resume from"
			}
		},
		"required": ["description", "prompt", "subagentType"]
	}`)
}

func (t *TaskTool) Execute(ctx context.Context, input json.RawMessage, toolCtx *Context) (*Result, error) {
	var params TaskInput
	if err := json.Unmarshal(input, &params); err != nil {
		return nil, fmt.Errorf("invalid input: %w", err)
	}

	// Validate required fields
	if params.Description == "" {
		return nil, fmt.Errorf("description is required")
	}
	if params.Prompt == "" {
		return nil, fmt.Errorf("prompt is required")
	}
	if params.SubagentType == "" {
		return nil, fmt.Errorf("subagentType is required")
	}

	// Get subagent configuration
	subagent, err := t.agentRegistry.Get(params.SubagentType)
	if err != nil {
		// Try with lowercase
		subagent, err = t.agentRegistry.Get(params.SubagentType)
		if err != nil {
			return nil, fmt.Errorf("unknown subagent type: %s. Available types: general, explore, plan", params.SubagentType)
		}
	}

	// Verify subagent mode
	if !subagent.IsSubagent() {
		return nil, fmt.Errorf("agent %s cannot be used as subagent (mode: %s)", params.SubagentType, subagent.Mode)
	}

	// Update metadata
	toolCtx.SetMetadata(params.Description, map[string]any{
		"subagent": params.SubagentType,
		"status":   "starting",
	})

	// If no executor is set, return a placeholder result
	if t.executor == nil {
		return &Result{
			Title:  fmt.Sprintf("Task: %s", params.Description),
			Output: fmt.Sprintf("[Subtask execution not configured]\n\nAgent: %s\nPrompt: %s", params.SubagentType, params.Prompt),
			Metadata: map[string]any{
				"subagent":    params.SubagentType,
				"status":      "skipped",
				"description": params.Description,
			},
		}, nil
	}

	// Execute subtask
	opts := TaskOptions{
		Model:       params.Model,
		ResumeFrom:  params.Resume,
		Description: params.Description,
	}

	result, err := t.executor.ExecuteSubtask(ctx, toolCtx.SessionID, params.SubagentType, params.Prompt, opts)
	if err != nil {
		return &Result{
			Title:  fmt.Sprintf("Subtask failed: %s", params.Description),
			Output: fmt.Sprintf("Error: %s", err.Error()),
			Metadata: map[string]any{
				"subagent": params.SubagentType,
				"status":   "failed",
				"error":    err.Error(),
			},
		}, nil
	}

	metadata := map[string]any{
		"subagent": params.SubagentType,
		"status":   "completed",
	}
	if result.SessionID != "" {
		metadata["sessionID"] = result.SessionID
	}
	if result.AgentID != "" {
		metadata["agentID"] = result.AgentID
	}
	if result.Metadata != nil {
		for k, v := range result.Metadata {
			metadata[k] = v
		}
	}

	return &Result{
		Title:    fmt.Sprintf("Completed: %s", params.Description),
		Output:   result.Output,
		Metadata: metadata,
	}, nil
}

func (t *TaskTool) EinoTool() einotool.InvokableTool {
	return &einoToolWrapper{tool: t}
}

// GetAvailableAgents returns a list of available subagent types.
func (t *TaskTool) GetAvailableAgents() []string {
	agents := t.agentRegistry.ListSubagents()
	names := make([]string, len(agents))
	for i, a := range agents {
		names[i] = a.Name
	}
	return names
}

// GetAgentDescription returns the description of a specific agent.
func (t *TaskTool) GetAgentDescription(name string) (string, error) {
	ag, err := t.agentRegistry.Get(name)
	if err != nil {
		return "", err
	}
	return ag.Description, nil
}

File: packages/kiana-v6/src/tool/task.ts
import { z } from "zod"
import { defineTool, ToolContext } from "./tool.js"

const DESCRIPTION = `Launch a new agent to handle complex, multi-step tasks autonomously.

Available agent types:
- general-purpose: General-purpose agent for researching complex questions, searching for code, and executing multi-step tasks.
- explore: Fast agent specialized for exploring codebases. Use this when you need to quickly find files by patterns, search code for keywords, or answer questions about the codebase.

When using the Task tool, you must specify a subagent_type parameter to select which agent type to use.

When to use the Task tool:
- When you are instructed to execute custom slash commands
- When exploring large codebases
- When tasks require multiple rounds of searching and analysis

When NOT to use the Task tool:
- If you want to read a specific file path, use the Read or Glob tool instead
- If you are searching for a specific class definition, use the Glob tool instead
- If you are searching for code within a specific file or set of 2-3 files, use the Read tool instead

Usage notes:
1. Launch multiple agents concurrently whenever possible for maximum performance
2. When the agent is done, it will return a single message back to you
3. Each agent invocation is stateless - provide a highly detailed task description
4. The agent's outputs should generally be trusted
5. Clearly tell the agent whether you expect it to write code or just to do research`

// Agent configuration for subagents
interface AgentConfig {
  name: string
  description: string
  systemPrompt: string
}

const AGENTS: Record<string, AgentConfig> = {
  "general-purpose": {
    name: "general-purpose",
    description: "General-purpose agent for researching complex questions and executing multi-step tasks",
    systemPrompt: `You are a helpful coding assistant running as a subagent. Complete the task autonomously and return a concise summary of your findings or actions.

When searching for code or files:
- Use Glob to find files by patterns
- Use Grep to search file contents
- Use Read to examine specific files

Be thorough but efficient. Return only the essential information needed to answer the question or complete the task.`,
  },
  explore: {
    name: "explore",
    description: "Fast agent specialized for exploring codebases",
    systemPrompt: `You are a fast code exploration agent. Your job is to quickly find files and code patterns in the codebase.

Use these tools efficiently:
- Glob: Find files by pattern (e.g., "**/*.ts", "src/**/*.tsx")
- Grep: Search file contents for keywords or patterns
- Read: Examine specific files when needed

Be quick and focused. Return a concise summary of what you found.`,
  },
}

// Subagent session executor type - will be injected by the session manager
export type SubagentExecutor = (params: {
  sessionID: string
  parentSessionID: string
  prompt: string
  agentConfig: AgentConfig
  agentType: string
  tools: Record<string, boolean>
  depth: number // nesting level (1 = direct subagent, 2 = subagent of subagent)
}) => Promise<{
  output: string
  sessionID: string
}>

// This will be set by the session manager
let subagentExecutor: SubagentExecutor | null = null

export function setSubagentExecutor(executor: SubagentExecutor) {
  subagentExecutor = executor
}

export function getSubagentExecutor(): SubagentExecutor | null {
  return subagentExecutor
}

export function getAgentConfig(agentType: string): AgentConfig | undefined {
  return AGENTS[agentType]
}

export function getAvailableAgentTypes(): string[] {
  return Object.keys(AGENTS)
}

export const taskTool = defineTool("task", {
  description: DESCRIPTION,
  parameters: z.object({
    description: z.string().describe("A short (3-5 words) description of the task"),
    prompt: z.string().describe("The task for the agent to perform"),
    subagent_type: z.string().describe("The type of specialized agent to use for this task"),
    session_id: z.string().describe("Existing Task session to continue").optional(),
  }),
  async execute(params, ctx) {
    const agent = AGENTS[params.subagent_type]
    if (!agent) {
      const availableAgents = Object.keys(AGENTS).join(", ")
      throw new Error(
        `Unknown agent type: ${params.subagent_type}. Available agents: ${availableAgents}`
      )
    }

    // Update metadata to show task in progress
    ctx.metadata({
      title: params.description,
      metadata: {
        agentType: params.subagent_type,
        status: "running",
      },
    })

    if (!subagentExecutor) {
      // Fallback: Return a message indicating subagent execution is not available
      // This happens when the task tool is used without a proper session manager
      return {
        title: params.description,
        metadata: {
          agentType: params.subagent_type,
          status: "skipped",
        },
        output: `Subagent execution is not available in this context. Task: ${params.prompt}`,
      }
    }

    try {
      const result = await subagentExecutor({
        sessionID: params.session_id || `task-${Date.now()}`,
        parentSessionID: ctx.sessionID,
        prompt: params.prompt,
        agentConfig: agent,
        agentType: params.subagent_type,
        tools: {
          todowrite: false,
          todoread: false,
          task: false, // Prevent recursive task calls
        },
        depth: 1, // Direct subagent call from parent
      })

      const output =
        result.output +
        "\n\n" +
        ["<task_metadata>", `session_id: ${result.sessionID}`, "</task_metadata>"].join("\n")

      return {
        title: params.description,
        metadata: {
          sessionId: result.sessionID,
          status: "completed",
        },
        output,
      }
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : String(error)
      return {
        title: params.description,
        metadata: {
          agentType: params.subagent_type,
          status: "error",
          error: errorMessage,
        },
        output: `Task failed: ${errorMessage}`,
      }
    }
  },
})
